launch_ros/launch_ros/utilities/lifecycle_event_manager.py

```diff
@@ -0,0 +1,142 @@
+              "/<name>/transition_event" topic, indicating the lifecycle
+              node represented by this action changed state
+
+        This utility also handles some events related to lifecycle:
+
+        - :class:`launch.events.lifecycle.ChangeState`
+
+          - this event can be targeted to a single lifecycle node, or more than
+            one, or even all lifecycle nodes, and it requests the targeted nodes
+            to change state, see its documentation for more details.
+
+        :param name: The name of the lifecycle node.
+        """
+        self.__logger = launch.logging.get_logger(__name__)
+        self.__node_name = name
+        self.__rclpy_subscription = None
+        self.__rclpy_change_state_client = None
+
+    @property
+    def node_name(self):
+        return self.__node_name
+
+    def _on_transition_event(self, context, msg):
+        try:
+            event = StateTransition(action=self, msg=msg)
+            context.asyncio_loop.call_soon_threadsafe(lambda: context.emit_event_sync(event))
+        except Exception as exc:
+            self.__logger.error(
+                "Exception in handling of 'lifecycle.msg.TransitionEvent': {}".format(exc))
+
+    def _call_change_state(self, request, context: launch.LaunchContext):
+        while not self.__rclpy_change_state_client.wait_for_service(timeout_sec=1.0):
+            if context.is_shutdown:
+                self.__logger.warning(
+                    "Abandoning wait for the '{}' service, due to shutdown.".format(
+                        self.__rclpy_change_state_client.srv_name),
+                )
+                return
+
+        # Asynchronously wait so that we can periodically check for shutdown.
+        event = threading.Event()
+
+        def unblock(future):
+            nonlocal event
+            event.set()
+
+        response_future = self.__rclpy_change_state_client.call_async(request)
+        response_future.add_done_callback(unblock)
+
+        while not event.wait(1.0):
+            if context.is_shutdown:
+                self.__logger.warning(
+                    "Abandoning wait for the '{}' service response, due to shutdown.".format(
+                        self.__rclpy_change_state_client.srv_name),
+                )
+                response_future.cancel()
+                return
+
+        if response_future.exception() is not None:
+            raise response_future.exception()
+        response = response_future.result()
+
+        if not response.success:
+            self.__logger.error(
+                "Failed to make transition '{}' for LifecycleNode '{}'".format(
+                    ChangeState.valid_transitions[request.transition.id],
+                    self.__node_name,
+                )
+            )
+
+    def _on_change_state_event(self, context: launch.LaunchContext) -> None:
+        typed_event = cast(ChangeState, context.locals.event)
+        if not typed_event.lifecycle_node_matcher(self):
```

Comment:
Again, if the user emits their own `ChangeState` pointing to their `LifecycleNode` action, then this will not match it right?

Reply:
This is copy+pasted from the lifecycle node abstracted into a manager to share with composition node. I'd prefer not to make functional changes in this PR if that's OK with you. This has been the state of things previously and just reorganized to be shared among multiple node types

But w.r.t. this comment: No, actually! The subscription keeps them aligned: https://github.com/ros2/launch_ros/blob/84c44de2ab85eb03e55ddcd6491a79eceba38482/launch_ros/launch_ros/utilities/lifecycle_event_manager.py#L127-L131

Reply:
> This is copy+pasted from the lifecycle node abstracted into a manager to share with composition node. I'd prefer not to make functional changes in this PR if that's OK with you. This has been the state of things previously and just reorganized to be shared among multiple node types

I get where you're coming from, but in this case your copy-paste is into a different class with a different type where you also use `self`, so you're essentially changing the behavior even though it's a copy-paste because of that.

I don't see  how the subscriptions keep them aligned? The `typed_event.lifecycle_node_matcher` is user defined and may not look at the subscription at all. Plus the matcher explicitly expects a `LifecycleNode` here not the type of self in your copy-pasted code:

https://github.com/ros2/launch_ros/blob/f9ab9371824d470bca47602bc7d277eb2217c90d/launch_ros/launch_ros/events/lifecycle/change_state.py#L92

Here's an example that won't likely work:

https://github.com/ros2/launch_ros/blob/f9ab9371824d470bca47602bc7d277eb2217c90d/launch_ros/examples/lifecycle_pub_sub_launch.py#L50

That is expecting the actions objects to match...

Reply:
I override the `__eq__` operator so that the lifecycle node object is used for those comparisons :-) 

```
    def __eq__(self, other):
        return self.__lifecycle_node == other
```

After the changes in the other comments, I tried all of the launch examples again and they all work (including setting autostart to false, true, and removed): 

Node via `pub_sub_launch.py`:
* No autostart, gives error if one is given

Lifecycle node via `lifecycle_pub_sub_launch.py` with external triggers:
* The registered event handlers and event emiting works properly, nodes come up as expected 
* I think this was your concern, this does work. 

Lifecycle nodes via `lifecycle_autostart_pub_sub_launch.py` with internal autostart:
* Autostart = true, autostarts and pub/sub is logged. Autostart logging is seen.
* Autostart=false, system comes up, no pub/sub logging. Autostart logging is not seen.
* Autostart not set, system comes up, no pub/sub logging. Autostart logging is not seen.

Lifecycle component nodes via `lifecycle_component_autostart_pub_sub_launch.py`:
* Autostart=true, Autostart logging is seen.
* Autostart=false, Autostart logging is not seen.
* Autostart not set, Autostart logging is not seen.

Note for the last one `lifecycle_component_autostart_pub_sub_launch.py`, the components themselves are not lifecycle enabled, so the logging is used to validate that the autostart keyword is parsed and will setup the transition. They end up publishing / subscribing regardless of autostart since they're not actually lifecycle nodes at all. Unfortunately the [lifecycle listener / talker demos](https://github.com/ros2/demos/blob/rolling/lifecycle/src/lifecycle_listener.cpp) are not components for me to use, so I have to use the components that we have. I did test and do debug logging to make sure this is correct and I'm confident that it is.

Reply:
> I override the __eq__ operator so that the lifecycle node object is used for those comparisons :-)

That's kind of my point, which is that the type annotations are indicating that you should expect something different. It just so happens that the matchers we use only do `==`, but a user sees it takes a `LifecycleNode` and may think they can call anything available on that class in their matcher, but that's not the case because you're passing a different class and in one case making it appear to be correct. It would be better to just pass the correct class, or have the new class actually inherit from `LifecycleNode`, or actually change the API of the callback for matchers so the false expectation is avoided.

My preference is for one of the first two options honestly. Trying to pass off this new class as if it were a `LifecycleNode` seems a bit too clever to me. It's gonna be surprising and confusing to someone at some point, in my opinion.

Reply:
I did this to avoid making other changes to matchers. I actually did this originally but backed that out because the `launch.events` matcher [1] compares the nodes rather than any specific attribute. The existing [lifecycle pub/sub launch example](https://github.com/ros2/launch_ros/blob/rolling/launch_ros/examples/lifecycle_pub_sub_launch.py#L50) uses that matcher and would break the behavior if `==` didn't match. Since `launch.events` is pretty core, I didn't want to introduce any change that would break that behavior, even though the [`launch_ros.events` versions work perfectly for us](https://github.com/ros2/launch_ros/blob/rolling/launch_ros/launch_ros/events/matchers/matches_node_name.py#L25-L29 [2]. The launch_ros version honestly seem better to me by being less fragile based on attributes rather than comparing full objects.  However anyone using the `launch.events` version rather than the `launch_ros.events` version would see a break, so I implemented the `__eq__` to make them both continue to work as expected. 

[1] https://github.com/ros2/launch/blob/rolling/launch/launch/events/matchers.py#L22-L24
[2] https://github.com/ros2/launch_ros/blob/rolling/launch_ros/launch_ros/events/matchers/matches_node_name.py#L25-L29

> It would be better to just pass the correct class, or have the new class actually inherit from LifecycleNode ... My preference is for one of the first two options honestly. 

I'm always open for critique - I don't really understand the suggestion though, can you clarify your vision for: "pass the right class"?  

I odn't think the new class is good since it would make users have to use a `LifecycleAutostartingNode` object just to obtain the `autostart` field, which is unnatural and seems confusing. Plus, it would require me to duplicate code. I'm on board with the semantic difference between `ComposableLifecycleNode` and `ComposableNode`, but having a second `LifecycleNode` for a feature that is intrinsically linked to lifecycle nodes is strange.   

> Trying to pass off this new class as if it were a LifecycleNode seems a bit too clever to me. It's gonna be surprising and confusing to someone at some point, in my opinion.

Pending on understanding your point about "pass the right class", I think what is currently here is a better solution. Its a little clever, but not really much so, I don't think. Especially if documented (which I can improve) and keeps all matchers working

Reply:
>  I don't really understand the suggestion though, can you clarify your vision for: "pass the right class"?

You already have a new class, `LifecycleEventManager`, which you're passing on this line as `self`. My point was that you should either pass `self. __lifecycle_node` here or have `LifecycleEventManager` inherit from `LifecycleNode` since that is the type that the matcher is expecting. That way you don't need to override the `==` behavior and the matcher is getting what it expects and future changes that might do more than `==` will continue to work.

Reply:
Naively, this is a concrete change that would satisfy me here:

```suggestion
        if not typed_event.lifecycle_node_matcher(self. __lifecycle_node):
```

Though that might not work for some other reason, in which case you might have to do the inheritance.

Reply:
Well, I guess not, because if there's something in the `LifecycleEventManager` class that you need in the matcher, you could just do `lifecycle_node.__lifecycle_event_manager` (which btw there should probably be an accessor for the event manager from the lifecycle node).

Reply:
> You already have a new class, LifecycleEventManager, which you're passing on this line as self. My point was that you should either pass self.__lifecycle_node here 

This is another obvious simplicity from heritage decisions where this started by only passing in the lifecycle node's name, not the object itself. Now that we have the object, that is exactly the right thing to do for the case of `_on_change_state_event`. I'll do that in a commit since there's absolutely no good reason not to do that. 

However, that doesn't address the issue of the `launch.event` matcher which uses the `==` operator - which can be defined externally by a user. [For example in the `lifecycle_pub_sub_launch.py` demo example](https://github.com/ros2/launch_ros/blob/rolling/launch_ros/examples/lifecycle_pub_sub_launch.py#L50). I don't actually think (?) that deriving from the lifecycle node class would solve that issue either, since the `LifecycleEventManager` would then be of type `LifecycleNode`, but not the same instance of it. 

Edit: Actually, I just found a way around that. I can just pass in the lifecycle node into the state transition too `StateTransition(action=self.__lifecycle_node, msg=msg)`. 

This is done :sunglasses: 

Personally, I think the `launch.event` matcher is extremely fragile. Checking that specific instances match 

 > which btw there should probably be an accessor for the event manager from the lifecycle node

ACKed and done

Reply:
> Personally, I think the launch.event matcher is extremely fragile. I feel like having the actions be required to have a name field which could be autogenerated as a hash if not provided would be better.

Not sure what you mean, but for XML having "named" instances of actions would be necessary/useful. But what you're describing (with autogeneration or hashing), isn't that just `id(action)`? Which btw is what `==` is checking when you call it by default:

https://docs.python.org/3/library/functions.html#id

Reply:
Huh, ok!